idempotency_header_middleware/backends/redis.py

```diff
@@ -55,12 +56,20 @@ async def store_response_data(self, idempotency_key: str, payload: dict, status_
 
     async def store_idempotency_key(self, idempotency_key: str) -> bool:
         """
-        Store an idempotency key header value in a set.
+        Store an idempotency key header value in a sortedset.
         """
-        return not bool(await self.redis.sadd(self.KEYS_KEY, idempotency_key))
+        return not bool(await self.redis.zadd(self.KEYS_KEY, {idempotency_key: time.time() + self.expiry},))
 
     async def clear_idempotency_key(self, idempotency_key: str) -> None:
         """
         Remove an idempotency header value from the set.
         """
-        await self.redis.srem(self.KEYS_KEY, idempotency_key)
+        await self.redis.zrem(self.KEYS_KEY, idempotency_key)
+
+    async def expire_idempotency_keys(self) -> None:
+        """
+        Remove any expired idempotency keys to avoid returning 409s
+        after the response expires.
+        """
+        if self.expiry:
+            await self.redis.zremrangebyscore(self.KEYS_KEY, '-inf', time.time())
```

Comment:
I mentioned it in the main PR thread, but here's where it might make sense to instead store each idempotency key as its own key in Redis and avoid the potentially expensive [`ZREMRANGEBYSCORE`](https://redis.io/commands/zremrangebyscore/) call entirely.

Reply:
I'm not very concerned with this being a slow operation, as a slow redis operation is probably much faster for this type of use-case than making just one additional round-trip to the cache. At least that's what I'd expect.

If we wanted to store them separately and save the round-trips, I suppose we could use pipelining or a Lua script to perform the operations in one go. I'll leave that up to you 🙂